pkg/datasource/data.go

```diff
@@ -0,0 +1,333 @@
+// Copyright 2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datasource
+
+import (
+	"fmt"
+	"io"
+	"reflect"
+	"slices"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+)
+
+type (
+	data  api.GadgetData
+	field api.Field
+)
+
+func (*data) private() {}
+
+func (f *field) ReflectType() reflect.Type {
+	switch Kind(f.Type) {
+	default:
+		return nil
+	case Int8:
+		return reflect.TypeOf(int8(0))
+	case Int16:
+		return reflect.TypeOf(int16(0))
+	case Int32:
+		return reflect.TypeOf(int32(0))
+	case Int64:
+		return reflect.TypeOf(int64(0))
+	case Uint8:
+		return reflect.TypeOf(uint8(0))
+	case Uint16:
+		return reflect.TypeOf(uint16(0))
+	case Uint32:
+		return reflect.TypeOf(uint32(0))
+	case Uint64:
+		return reflect.TypeOf(uint64(0))
+	case Float32:
+		return reflect.TypeOf(float32(0))
+	case Float64:
+		return reflect.TypeOf(float64(0))
+	case Bool:
+		return reflect.TypeOf(false)
+	case String:
+		return reflect.TypeOf("")
+	}
+}
+
+type dataSource struct {
+	name string
+	id   uint32
+
+	dType Type
+	dPool sync.Pool
+
+	// keeps information on registered fields
+	fields   []*field
+	fieldMap map[string]*field
+
+	payloadCount uint32
+
+	requestedFields map[string]bool
+
+	subscriptions []*subscription
+
+	requested bool
+
+	lock sync.RWMutex
+}
+
+func New(t Type, name string) DataSource {
+	ds := &dataSource{
+		name:            name,
+		dType:           t,
+		requestedFields: make(map[string]bool),
+		fieldMap:        make(map[string]*field),
+	}
+	ds.registerPool()
+	return ds
+}
+
+func NewFromAPI(in *api.DataSource) (DataSource, error) {
+	ds := &dataSource{
+		name:            in.Name,
+		dType:           Type(in.Type),
+		requestedFields: make(map[string]bool),
+		fieldMap:        make(map[string]*field),
+	}
+	for _, f := range in.Fields {
+		ds.fields = append(ds.fields, (*field)(f))
+		ds.fieldMap[f.Name] = (*field)(f)
+	}
+	ds.registerPool()
+	return ds, nil
```

Comment:
> > to avoid refactoring the callers
> 
> `gopatch` can handle this. You are already refactoring, so let's merge correct code and only what we need.

I don't agree with this. Why to have to change this in the future if this is planned to return errors later on? It's better to define the right function signature from the beginning.

Reply:
From my understanding we agreed to keep it this way to not have to change the function signature later on and have to edit all the callers.